pkg/binding/binding.go

```diff
@@ -0,0 +1,149 @@
+		flagsBackend:       backendpkg.NewFlagsBackend(),
+		interactiveBackend: backendpkg.NewInteractiveBackend(askerClient),
+		kubernetesClient:   kubernetesClient,
+	}
+}
+
+// GetFlags gets the flag specific to add binding operation so that it can correctly decide on the backend to be used
+// It ignores all the flags except the ones specific to add binding operation, for e.g. verbosity flag
+func (o *BindingClient) GetFlags(flags map[string]string) map[string]string {
+	bindingFlags := map[string]string{}
+	for flag, value := range flags {
+		if flag == backendpkg.FLAG_NAME || flag == backendpkg.FLAG_SERVICE || flag == backendpkg.FLAG_BIND_AS_FILES {
+			bindingFlags[flag] = value
+		}
+	}
+	return bindingFlags
+}
+
+// Validate calls Validate method of the adequate backend
+func (o *BindingClient) Validate(flags map[string]string) error {
+	var backend backendpkg.AddBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.Validate(flags)
+}
+
+func (o *BindingClient) SelectServiceInstance(flags map[string]string, serviceMap map[string]unstructured.Unstructured) (string, error) {
+	var backend backendpkg.AddBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.SelectServiceInstance(flags[backendpkg.FLAG_SERVICE], serviceMap)
+}
+
+func (o *BindingClient) AskBindingName(serviceName, componentName string, flags map[string]string) (string, error) {
+	var backend backendpkg.AddBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	defaultBindingName := fmt.Sprintf("%v-%v", componentName, serviceName)
+	return backend.AskBindingName(defaultBindingName, flags)
+}
+
+func (o *BindingClient) AskBindAsFiles(flags map[string]string) (bool, error) {
+	var backend backendpkg.AddBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.AskBindAsFiles(flags)
+}
+
+func (o *BindingClient) AddBinding(bindingName string, bindAsFiles bool, unstructuredService unstructured.Unstructured, obj parser.DevfileObj, componentContext string) error {
+	service, err := o.kubernetesClient.NewServiceBindingServiceObject(unstructuredService, bindingName)
+	if err != nil {
+		return err
+	}
+
+	deploymentName := fmt.Sprintf("%s-app", obj.GetMetadataName())
+	deploymentGVR, err := o.kubernetesClient.GetDeploymentAPIVersion()
+	if err != nil {
+		return err
+	}
+
+	serviceBinding := o.kubernetesClient.NewServiceBindingObject(bindingName, bindAsFiles, deploymentName, deploymentGVR, []sboApi.Mapping{}, []sboApi.Service{service})
+
+	// Note: we cannot directly marshal the serviceBinding object to yaml because it doesn't do that in the correct k8s manifest format
+	serviceBindingUnstructured, err := kclient.ConvertK8sResourceToUnstructured(serviceBinding)
+	if err != nil {
+		return err
+	}
```

Comment:
You could use `ToUnstructuredAsGVK` defined [here](https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/converter/unstructured.go) so you don't need to build the `TypeMeta` in `NewServiceBindingObject`. And you could use the gvk of ServiceBinding defined in kclient/binding

Reply:
Or you can use the GroupVersionResource and GroupVersionKind defined in https://github.com/redhat-developer/service-binding-operator/blob/master/apis/binding/v1alpha1/groupversion_info.go

Reply:
I am not very sure about this. The reason I added NewServiceBindingObject request to is to ensure I return a complete SB object. If I remove the `TypeMeta`, and if this method is called from some place, chances are it will not be called with `ToUnstructuredAsGVK`. Don't you think it will fail in the case where it's called with some other method?

Reply:
Yes, the problem is that the Kind and APIVersion won't be added without `ToUnstructuredAsGVK`. Let's keep as this